### PR TITLE
Refactor and unify user dispatcher apis

### PR DESCRIPTION
**Pull request by pengzhout**
This commit unified the dispatcher apis to make it more clear to callers and it expose more
flexible interfaces. Meanwhile it provide better error handling, caller do not need care about cleaning of dispatch context when errors happen.

details please see each commit message

### CODE REVIEW

File: src/backend/cdb/cdbtm.c [hunk 1]
```diff
@@ -2373,15 +2365,10 @@ dispatchDtxCommand(const char *cmd, bool withSnapshot, bool raiseError)
...
 	}
 
-	/* discard the errbuf text */
-	pfree(errbuf.data);
-
 	/* Now we clean up the results array. */
 	for (i = 0; i < resultCount; i++)
 		PQclear(results[i]);
 	free(results);
-
```
Comment (xiong-gang) [inline src/backend/cdb/cdbtm.c]: if numOfFailed >0 should report a error

File: src/backend/cdb/dispatcher/cdbdisp_query.c [hunk 2]
```diff
@@ -1581,6 +1496,165 @@ buildSliceIndexGangIdMap(SliceVec *sliceVec, int numSlices, int numTotalSlices)
...
+	 */
+	results = cdbdisp_getDispatchResults(ds, &errorMsg);
+	
+	/*
+	 * If QEs have errors, throw it up
+	 */
+	if (!results && errorMsg)
+	{
+		cdbdisp_destroyDispatcherState(ds);	
+		elog(ERROR, "cdbdisp_dispatchStatementNode: One or more QEs got Errors: %s",
```
Comment (xiong-gang) [inline src/backend/cdb/dispatcher/cdbdisp_query.c]: should change elog to ereport so that dev can quick locate which line report the error.

File: src/backend/cdb/dispatcher/cdbdisp.c [hunk 3]
```diff
@@ -179,211 +179,6 @@ cdbdisp_getDispatchResults(struct CdbDispatcherState *ds, StringInfoData **error
...
-		if (qderrtext
-			&& strcmp(qderrtext, CDB_MOTION_LOST_CONTACT_STRING) == 0)
-			qd_lost_flag = true;
-
-		if (ds->primaryResults && ds->primaryResults->errcode)
-		{
-			if (qd_lost_flag
-				&& ds->primaryResults->errcode == ERRCODE_GP_INTERCONNECTION_ERROR)
-				useQeError = true;
-			else if (ds->primaryResults->errcode != ERRCODE_GP_INTERCONNECTION_ERROR)
```
Comment (pengzhout) [inline src/backend/cdb/dispatcher/cdbdisp.c]: handleError should be reserved when dispatching is sync mode (cdbdisp_dispatchPlan).

File: src/include/cdb/cdbdisp_query.h [hunk 4]
```diff
@@ -177,4 +110,60 @@ CdbExecCommandOnQEsExpectResults(const char* strCommand,
...
+
+#define CdbDoUtility_COE_2PC_SNAPSHOT(stmt, debugCaller)\
+	CdbExecUtilityStatementOnQEs(stmt, true, true, true, debugCaller)
+
+#define CdbDoUtility_2PC_SNAPSHOT(stmt, debugCaller)\
+	CdbExecUtilityStatementOnQEs(stmt, false, true, true, debugCaller)
+
+#define CdbDoUtilityV1_COE_2PC_SNAPSHOT(stmt, numResultPtr, debugCaller)\
+	CdbExecUtilityStatementOnQEsExpectResults(stmt, true, true, true, numResultPtr, debugCaller)
+/*
```
Comment (hlinnaka) [inline src/include/cdb/cdbdisp_query.h]: I find these macros a bit hard to read. I agree that having three boolean arguments is confusing too, so I can see why you added the macros.

How about using a 'flags' argument? Something along the lines of:

```
#define EUS_CANCEL_ON_ERROR 0x1
#define EUS_NEED_TWO_PHASE 0x2
#define EUS_WITH_SNAPSHOT  0x4

extern void CdbExecUtilityStatementOnQEs(Node *stmt, int flags);
```

IMHO that makes the call sites much more clear.

Can we just remove the unused 'debugCaller' argument?

File: src/backend/cdb/dispatcher/cdbdisp.c [hunk 5]
```diff
@@ -141,6 +141,44 @@ CdbCheckDispatchResult(struct CdbDispatcherState *ds,
...
+	CdbCheckDispatchResult(ds, DISPATCH_WAIT_NONE);
+
+	/* check if any error reported */
+	errorcode = ds->primaryResults->errcode;
+
+	if (errorcode)
+	{
+		errorMsgBuf = palloc0(sizeof(StringInfoData));
+		initStringInfo(errorMsgBuf);
+		cdbdisp_dumpDispatchResults(ds->primaryResults, errorMsgBuf, false);
```
Comment (hlinnaka) [inline src/backend/cdb/dispatcher/cdbdisp.c]: Should use "makeStringInfo()".

Or actually, it's a bit unusual that cdbdisp_getDispatchResults() returns a "StringInfoData *" in the first place. I would expect it to append to a caller-supplied StringInfo, if it's expected that the caller will somehow manipulate the buffer further. That doesn't seem likely though. The caller is probably going to just print out the error. In that case, it would seem better for this function to return a plain string containing the error message.

**Comment by hlinnaka** [comment]
Thanks, I like these refactorings a lot at a high level.

I'd suggest squashing the first commit that just adds new functions, with the next commit that uses those new functions. It's quite arbitrary to split those into separate commits, as neither of them make any sense without the other.

This patch set rewrites many places that did a custom ereport() on error, with the dbDoCommand_*() functions/macros, which throws a more generic error. That certainly makes the code simpler, but I think you lose some details from the error message of what exactly went wrong. Is that OK? For example, the error message used to be "pg_highest_oid error (gathered %d results from cmd '%s'". I'm not sure what you'd get for the same error now, but surely less details. Should some of the callers install an error context callback to add a CONTEXT line to those messages, perhaps?

File: src/include/cdb/cdbdisp_query.h (hunk 4 above)
Comment (pengzhout) [inline src/include/cdb/cdbdisp_query.h]: > How about using a 'flags' argument? Something along the lines of:
>
> #define EUS_CANCEL_ON_ERROR 0x1
> #define EUS_NEED_TWO_PHASE 0x2
> #define EUS_WITH_SNAPSHOT  0x4
>
> extern void CdbExecUtilityStatementOnQEs(Node *stmt, int flags);
>
> IMHO that makes the call sites much more clear.
>
Yes, flags seem to be better , thanks.

**Comment by pengzhout** [comment]
>
> Thanks, I like these refactorings a lot at a high level.
>
> I'd suggest squashing the first commit that just adds new functions, with
> the next commit that uses those new functions. It's quite arbitrary to
> split those into separate commits, as neither of them make any sense
> without the other.
>
> This patch set rewrites many places that did a custom ereport() on error,
> with the dbDoCommand_*() functions/macros, which throws a more generic
> error. That certainly makes the code simpler, but I think you lose some
> details from the error message of what exactly went wrong. Is that OK? For
> example, the error message used to be "pg_highest_oid error (gathered %d
> results from cmd '%s'". I'm not sure what you'd get for the same error now,
> but surely less details. Should some of the callers install an error
> context callback to add a CONTEXT line to those messages, perhaps?
>
>
> This is should be a problem, the actual error details are not gone, but
the location is missed.  what about error out the command within
dbDoCommand_*() so we can recognize which command goes wrong, for example:

1.
 'select pg_highest_oid()' encounter dispatch error: details: xxxxxxxxxx
*VS*
pg_highest_oid error (gathered 2 results from cmd 'select
pg_highest_oid()') xxxxxxxxxxxxx

2.
 'select gid from pg_prepared_xacts encounter dispatch error: details:
xxxxxxxxxx'
*VS*
 Unable to collect list of prepared transactions from all segment
databases, xxxxxxxxxxxxx

3.
SAVEPOINT encounter dispatch error: details: xxxxxxxxx

*VS*
"Could not create a new savepoint (SAVEPOINT)"

"DTM error (gathered 2 results from cmd 'SAVEPOINT')"

"details: xxxxxxx"

**Comment by hlinnaka** [comment]
On 07/12/2016 06:16 AM, Tang Pengzhou wrote:
>> This patch set rewrites many places that did a custom ereport() on error,
>> with the dbDoCommand_*() functions/macros, which throws a more generic
>> error. That certainly makes the code simpler, but I think you lose some
>> details from the error message of what exactly went wrong. Is that OK? For
>> example, the error message used to be "pg_highest_oid error (gathered %d
>> results from cmd '%s'". I'm not sure what you'd get for the same error now,
>> but surely less details. Should some of the callers install an error
>> context callback to add a CONTEXT line to those messages, perhaps?
>
> This is should be a problem, the actual error details are not gone, but
> the location is missed.  what about error out the command within
> dbDoCommand_*() so we can recognize which command goes wrong, for example:
>
> 1.
>  'select pg_highest_oid()' encounter dispatch error: details: xxxxxxxxxx
> *VS*
> pg_highest_oid error (gathered 2 results from cmd 'select
> pg_highest_oid()') xxxxxxxxxxxxx
>
> 2.
>  'select gid from pg_prepared_xacts encounter dispatch error: details:
> xxxxxxxxxx'
> *VS*
>  Unable to collect list of prepared transactions from all segment
> databases, xxxxxxxxxxxxx
>
> 3.
> SAVEPOINT encounter dispatch error: details: xxxxxxxxx
>
> *VS*
> "Could not create a new savepoint (SAVEPOINT)"
>
> "DTM error (gathered 2 results from cmd 'SAVEPOINT')"
>
> "details: xxxxxxx"

Works for me. To follow the error message guidelines [1], it should be 
probably be spelled something along the lines of:

ERROR: could not dispatch command to segments
DETAIL: xxxxxxxxxx
CONTEXT: command "select pg_highest_oid()"


[1] https://www.postgresql.org/docs/devel/static/error-style-guide.html

- Heikki

**Comment by pengzhout** [comment]
>
>
>
> Works for me. To follow the error message guidelines [1], it should be
> probably be spelled something along the lines of:
>
> ERROR: could not dispatch command to segments
> DETAIL: xxxxxxxxxx
> CONTEXT: command "select pg_highest_oid()"
>
>
>
Thanks, fixed.

File: src/backend/cdb/cdbcopy.c [hunk 6]
```diff
@@ -177,32 +170,13 @@ cdbCopyStart(CdbCopy *c, char *copyCmd)
...
-			false /* in cursor */);
-	
-	cdbdisp_dispatchCommand(copyCmd, serializedQuerytree, serializedQuerytree_len, 
-								false 		/* cancelonError */, 
-								c->copy_in 	/* need2phase */, 
-								true 		/* withSnapshot */,
-								&ds);
+	CdbDispatchUtilityStatement((Node *)q->utilityStmt,
+								EUS_WITH_SNAPSHOT|
+								(c->copy_in ? EUS_NEED_TWO_PHASE : EUS_NONE),
```
Comment (xiong-gang) [inline src/backend/cdb/cdbcopy.c]: c->copy_in ? EUS_NEED_TWO_PHASE  |  EUS_WITH_SNAPSHOT :  EUS_NEED_TWO_PHASE

File: src/backend/cdb/cdbtm.c [hunk 7]
```diff
@@ -2250,32 +2250,32 @@ doDispatchDtxProtocolCommand(DtxProtocolCommand dtxProtocolCommand, int flags,
...
-	results = cdbdisp_dispatchDtxProtocolCommand(dtxProtocolCommand, flags,
+	cdb_pgresults = cdbdisp_dispatchDtxProtocolCommand(dtxProtocolCommand, flags,
 												 dtxProtocolCommandStr,
 												 gid, gxid,
-												 &errbuf, &resultCount, badGangs, direct,
+												 &errbuf, badGangs, direct,
 												 serializedDtxContextInfo, serializedDtxContextInfoLen);
 
-	if (errbuf.len > 0)
+	if (!cdb_pgresults && errbuf.len > 0)
```
Comment (xiong-gang) [inline src/backend/cdb/cdbtm.c]: || instead of && ?

File: src/backend/cdb/dispatcher/cdbdisp.c [hunk 8]
```diff
@@ -178,48 +206,22 @@ cdbdisp_finishCommand(struct CdbDispatcherState *ds,
...
 
-	/*
-	 * If no errors, free the CdbDispatchResults objects and return.
-	 */
-	if (ds->primaryResults)
-		errorcode = ds->primaryResults->errcode;
+	pr = cdbdisp_getDispatchResults(ds, &qeErrorMsg);
 
-	if (!errorcode)
+	if (!pr && qeErrorMsg.len > 0)
```
Comment (xiong-gang) [inline src/backend/cdb/dispatcher/cdbdisp.c]: || instead of &&

File: src/backend/cdb/dispatcher/cdbdisp.c [hunk 9]
```diff
@@ -178,48 +206,22 @@ cdbdisp_finishCommand(struct CdbDispatcherState *ds,
...
-		/*
-		 * Call the callback function to handle the results
-		 */
-		if (handle_results_callback != NULL)
-			handle_results_callback(ds->primaryResults, ctx);
-
 		cdbdisp_destroyDispatcherState(ds);
-		return;
-	}
+		ereport(ERROR, (errcode(ERRCODE_INTERNAL_ERROR),
```
Comment (xiong-gang) [inline src/backend/cdb/dispatcher/cdbdisp.c]: should we catch it and free qeErrorMsg?

File: src/backend/cdb/dispatcher/cdbdispatchresult.c [hunk 10]
```diff
@@ -801,16 +801,15 @@ cdbdisp_resultEnd(CdbDispatchResults * results, int sliceIndex)
...
+cdbdisp_returnResults(CdbDispatchResults * primaryResults)
 {
 	CdbDispatchResult *dispatchResult;
-	PGresult **resultSets = NULL;
 	int	nslots;
 	int	nresults = 0;
 	int	i;
-	int	totalResultCount = 0;
+
+	CdbPgResults *cdb_pgresults = (CdbPgResults*)palloc0(sizeof(CdbPgResults));
```
Comment (xiong-gang) [inline src/backend/cdb/dispatcher/cdbdispatchresult.c]: I would pass a CdbPgResults pointer as an argument, rather than malloc it inside.

File: src/backend/cdb/dispatcher/cdbdisp_dtx.c [hunk 11]
```diff
@@ -128,27 +127,41 @@ cdbdisp_dispatchDtxProtocolCommand(DtxProtocolCommand dtxProtocolCommand,
...
+		{
+			cdb_pgresults = cdbdisp_returnResults(pr);
+		}
 
-	cdbdisp_destroyDispatcherState((struct CdbDispatcherState *) &ds);
+		cdbdisp_destroyDispatcherState(&ds);
+	}
+	PG_CATCH();
+	{
+		cdbdisp_cancelDispatch(&ds);
```
Comment (xiong-gang) [inline src/backend/cdb/dispatcher/cdbdisp_dtx.c]: Should we cancel it?

File: src/backend/cdb/dispatcher/cdbdisp_query.c [hunk 12]
```diff
@@ -604,37 +620,46 @@ CdbDoCommand(const char *strCommand, bool cancelOnError, bool needTwoPhase)
...
-cdbdisp_dispatchUtilityStatement(struct Node *stmt,
-								 bool cancelOnError,
-								 bool needTwoPhase,
-								 bool withSnapshot,
-								 struct CdbDispatcherState *ds,
-								 char *debugCaller)
+CdbPgResults*
+CdbDispatchUtilityStatement(struct Node *stmt,
+							int flags,
+							bool returnPgResults)
```
Comment (xiong-gang) [inline src/backend/cdb/dispatcher/cdbdisp_query.c]: I would like to pass a pointer instead of a bool here.

File: src/backend/cdb/dispatcher/cdbdisp_query.c [hunk 13]
```diff
@@ -604,37 +620,46 @@ CdbDoCommand(const char *strCommand, bool cancelOnError, bool needTwoPhase)
...
-				  needTwoPhase, withSnapshot, false /* inCursor */ );
+	CdbDispatchResults* results = NULL;
+	StringInfoData qeErrorMsg;
 
-	initStringInfo(&buffer);
+	elog(((Debug_print_full_dtm || Debug_print_snapshot_dtm) ? LOG : DEBUG5),
+			"CdbDispatchUtilityStatement for command = '%s', cancelOnError = %s,\
+         needTwoPhase = %s, withSnapshot = %s",
+		 debug_query_string,
+		 "true",
```
Comment (xiong-gang) [inline src/backend/cdb/dispatcher/cdbdisp_query.c]: flags | EUS_NEED_CANCEL_ON_ERROR ? "true" : "false"

File: src/backend/cdb/dispatcher/cdbdisp_query.c [hunk 14]
```diff
@@ -660,13 +685,55 @@ cdbdisp_dispatchUtilityStatement(struct Node *stmt,
...
+
+	if (returnPgResults)
+	{
+		PG_TRY();
+		{
+			cdb_pgresults = cdbdisp_returnResults(results);
+		}
+		PG_CATCH();
+		{
+			cdbdisp_destroyDispatcherState(&ds);
```
Comment (xiong-gang) [inline src/backend/cdb/dispatcher/cdbdisp_query.c]: re-throw it

File: src/backend/cdb/dispatcher/cdbdispatchresult.c [hunk 15]
```diff
@@ -801,16 +801,15 @@ cdbdisp_resultEnd(CdbDispatchResults * results, int sliceIndex)
 	return &results->resultArray[si->resultEnd];
 }
 
-struct pg_result **
-cdbdisp_returnResults(CdbDispatchResults * primaryResults,
-					  StringInfo errmsgbuf, int *numresults)
+CdbPgResults*
```
Comment (xiong-gang) [inline src/backend/cdb/dispatcher/cdbdispatchresult.c]: Please add comments for this function, explain why don't handle error message here.

File: src/backend/commands/dbcommands.c [hunk 16]
```diff
@@ -93,8 +93,6 @@ static bool get_db_info(const char *name, LOCKMODE lockmode,
 static bool have_createdb_privilege(void);
 static bool check_db_file_conflict(Oid db_id);
 
-static void createdb_int(CreatedbStmt *stmt, CdbDispatcherState *ds);
```
Comment (xiong-gang) [inline src/backend/commands/dbcommands.c]: Please extract this as another commit.

File: src/backend/commands/indexcmds.c [hunk 17]
```diff
@@ -723,34 +727,10 @@ DefineIndex(RangeVar *heapRelation,
...
-			/* If dispatched, stop QEs and clean up after them. */
-			if (ds.primaryResults)
-				cdbdisp_handleError((struct CdbDispatcherState *)&ds);
-
-			PG_RE_THROW();
-			/* not reached */
-		}
-		PG_END_TRY();
+        CdbDispatchUtilityStatement((Node *) stmt,
+        								EUS_CANCEL_ON_ERROR|
```
Comment (xiong-gang) [inline src/backend/commands/indexcmds.c]: indention, i suggest put the "or" in one line since it's one argument and it's not that long, right?

File: src/backend/commands/tablecmds.c [hunk 18]
```diff
@@ -401,43 +399,6 @@ static char *alterTableCmdString(AlterTableType subtype);
 Oid
 DefineRelation(CreateStmt *stmt, char relkind, char relstorage)
 {
-	volatile struct CdbDispatcherState ds = {NULL, NULL};
```
Comment (xiong-gang) [inline src/backend/commands/tablecmds.c]: extract this as another commit. Same as defineindex

File: src/backend/commands/tablecmds.c [hunk 19]
```diff
@@ -836,13 +799,11 @@ DefineRelation_int(CreateStmt *stmt,
...
 		 */
-		cdbdisp_dispatchUtilityStatement((Node *)stmt,
-										 true,      /* cancelOnError */
-										 true,      /* startTransaction */
-										 true,      /* withSnapshot */
-										 ds,
-										 "DefineRelation_int");
-
+        CdbDispatchUtilityStatement((Node *) stmt,
+        								EUS_CANCEL_ON_ERROR|
```
Comment (xiong-gang) [inline src/backend/commands/tablecmds.c]: indention

File: src/backend/commands/tablecmds.c [hunk 20]
```diff
@@ -1173,87 +1133,64 @@ DefineExternalRelation(CreateExternalStmt *createExtStmt)
...
-            cdbdisp_handleError((struct CdbDispatcherState *)&ds);
 
-        /* Carry on with error handling. */
-		PG_RE_THROW();
+		/*
+		 * Dispatch the statement tree to all primary segdbs.
+		 * Doesn't wait for the QEs to finish execution.
+		 */
+        CdbDispatchUtilityStatement((Node *) createExtStmt,
+        								EUS_CANCEL_ON_ERROR|
```
Comment (xiong-gang) [inline src/backend/commands/tablecmds.c]: indention

File: src/backend/cdb/dispatcher/cdbdispatchresult.c [hunk 21]
```diff
@@ -822,9 +821,10 @@ cdbdisp_returnResults(CdbDispatchResults * primaryResults,
 		for (i = 0; i < primaryResults->resultCount; ++i)
```
Comment (xiong-gang) [inline src/backend/cdb/dispatcher/cdbdispatchresult.c]: we can do a little refactor here

File: src/backend/access/appendonly/appendonlywriter.c [hunk 22]
```diff
@@ -1315,45 +1312,43 @@ GetTotalTupleCountFromSegments(Oid aoRelid,
 	 */
 	GetUserIdAndContext(&save_userid, &save_secdefcxt);
 	SetUserIdAndContext(BOOTSTRAP_SUPERUSERID, true);
-	results = cdbdisp_dispatchRMCommand(sqlstmt.data, true,
-										&errbuf, &resultCount);
+
+	cdb_pgresults = CdbDispatchCommand(sqlstmt.data, EUS_WITH_SNAPSHOT, true);
+
 	/* Restore userid */
 	SetUserIdAndContext(save_userid, save_secdefcxt);
```
Comment (xiong-gang) [inline src/backend/access/appendonly/appendonlywriter.c]: Do we need to restore user id if dispatch goes wrong?

**Comment by pengzhout** [comment]
Thanks for reviews, closed and merged.
